### PR TITLE
wait for user network availability

### PR DESCRIPTION
**Pull request by epoberezkin**

### CODE REVIEW

**Review by dpwiz — DISMISSED** [review]

File: src/Simplex/Messaging/Agent.hs [hunk 1]
```diff
@@ -1267,6 +1277,7 @@ runSmpQueueMsgDelivery c@AgentClient {subQ} ConnData {connId} sq (Worker {doWork
         let mId = unId msgId
             ri' = maybe id updateRetryInterval2 msgRetryState ri
         withRetryLock2 ri' qLock $ \riState loop -> do
+          lift $ waitForUserNetwork c
```
Comment (spaced4ndy) [inline src/Simplex/Messaging/Agent.hs:1280]: why not before `beginAgentOperation c AOSndNetwork`, or at least after `endAgentOperation c AOMsgDelivery` before withRetryLock2? Not saying it's not correct, just wondering why it's not earlier

File: src/Simplex/Messaging/Agent/Client.hs [hunk 2]
```diff
@@ -581,6 +600,7 @@ resubscribeSMPSession c@AgentClient {smpSubWorkers} tSess =
       withRetryInterval ri $ \_ loop -> do
         pending <- atomically getPending
         forM_ (L.nonEmpty pending) $ \qs -> do
+          waitForUserNetwork c
```
Comment (spaced4ndy) [inline src/Simplex/Messaging/Agent/Client.hs:603]: add to ntf/xftp? currently it's only in resubscribeSMPSession and runSmpQueueMsgDelivery

File: src/Simplex/Messaging/Agent/Client.hs [hunk 3]
```diff
@@ -730,6 +750,26 @@ getClientConfig AgentClient {useNetworkConfig} cfgSel = do
   networkConfig <- readTVarIO useNetworkConfig
   pure cfg {networkConfig}
 
+waitForUserNetwork :: AgentClient -> AM' ()
+waitForUserNetwork AgentClient {userNetworkState} =
+  readTVarIO userNetworkState >>= \case
+    UNSOnline -> pure ()
+    UNSOffline {offlineDelay = d} ->
+      race (atomically $ unlessM online retry) (liftIO $ threadDelay' d) >>= \case
```
Comment (dpwiz) [inline src/Simplex/Messaging/Agent/Client.hs:758]: avoid spawning 4 threads with `registerDelay`
https://github.com/simplex-chat/simplexmq/pull/1086

File: src/Simplex/Messaging/Agent/Client.hs [hunk 4]
```diff
@@ -725,11 +760,39 @@ hostEvent :: forall v err msg. (ProtocolTypeI (ProtoType msg), ProtocolServerCli
...
   pure cfg {networkConfig}
 
+getNetworkConfig :: AgentClient -> STM NetworkConfig
+getNetworkConfig c = do
+  (slowCfg, fastCfg) <- readTVar (useNetworkConfig c)
+  UserNetworkState {networkType} <- readTVar (userNetworkState c)
+  pure $ case networkType of
+    UNCellular -> slowCfg
+    UNNone -> slowCfg
+    _ -> fastCfg
```
Comment (spaced4ndy) [inline src/Simplex/Messaging/Agent/Client.hs:775]:
```suggestion
    UNWifi -> fastCfg
    UNEthernet -> fastCfg
    UNOther -> fastCfg
```

**Pull request merged**